### PR TITLE
Copter: improve object avoidance using mini-fence

### PR DESCRIPTION
**Pull request by rmackay9**
This PR implements these changes:

- AP_Proximity gets "boundary points" which can be used by AC_Avoidance just as if it were a polygon fence
- AP_Proximity gets "ignore areas" so the sf40c can ignore up to 6 areas allowing the sf40c to ignore the vehicle legs if it's mounted on the bottom of the vehicle.
- AC_Avoidance modified to use the proximity sensor's boundary points
- Vehicle pre-arm check modified to show the angle to the closest object within 60cm of the vehicle (useful for setting up the ignore areas)

Note: these changes have been tested in SITL and in particular it improves object avoidance in situations where the vehicle is flying parallel to a wall.  The previous method allowed the vehicle to get infinitely close to a wall if the pilot pushes the pitch forward to make a vehicle fly towards a wall and then applies gentle consistent yaw.  This PR improves this so that the vehicle maintains the desired 2m distance from the wall.

Here's a picture of how the mini fence is built.  In short, the fence points are on the line segments between the sectors at a distance equal to the closest object in either adjacent segment.
avoidancemethods3

### CODE REVIEW

**Comment by dricketts** [comment]
@rmackay9 Have you considered merging this pull request first: https://github.com/ArduPilot/ardupilot/pull/4829? It might make sense to fix the fence avoidance logic before adding more sophisticated obstacle avoidance.

**Comment by rmackay9** [comment]
@dricketts, yes, you're right of course..

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -259,6 +259,43 @@ void NOINLINE Copter::send_rangefinder(mavlink_channel_t chan)
...
+    }
+
+    // send known distances
+    AP_Proximity::Proximity_Distance_Array dist_array;
+    if (g2.proximity.get_distances(dist_array)) {
+        for (uint8_t i=0; i<count_max; i++) {
+            mavlink_msg_distance_sensor_send(
+                chan,
+                AP_HAL::millis(),               //  time since system boot
+                (uint16_t)(g2.proximity.distance_min() * 100),  // minimum distance the sensor can measure in centimeters
```
Comment (muramura) [inline ArduCopter/GCS_Mavlink.cpp]: why c language cast used.
this source is c++.
c++ cast is static_cast  using , up quality.

Comment (OXINARF) [inline ArduCopter/GCS_Mavlink.cpp]: @muramura We use C casts almost everywhere and there's no rule saying not to use it and, as far as I know, there's no even a preference to use a C++ cast.

Comment (muramura) [inline ArduCopter/GCS_Mavlink.cpp]: @OXINARF Comment Thank you.
Whatever its form, cast should not be used.
I agree with that idea.

C language cast can force type conversion.
For example,
"Static_cast" of C ++ cast checks if the type specified at compile time is acceptable.
I think that it is better to use this check function.

**Comment by rmackay9** [comment]
Here's a flight test of this PR.  https://www.youtube.com/watch?v=1fSJNhRPBag
I added in two more changes to deal with issues I found:

- log the closest object which helps with debugging what's going wrong
- ensure the fence boundary is at least 60cm from the vehicle even if the lidar is reporting less than that distance.  This is important because otherwise the avoidance code simply ignores the entire polygon which means if a single sector returned a zero distance.

**Comment by rmackay9** [comment]
Merged!

**Pull request closed without merging**